### PR TITLE
V2 streaming

### PR DESCRIPTION
**Pull request by NassiHarel**
This change is [Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/892)

### CODE REVIEW

File: core/worker/lib/streaming/services/auto-scaler.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    async _election() {
+        const { childs, jobId } = this._jobData;
+        const data = { config: this._options, pipeline: this._pipeline, jobData: this._jobData, jobId };
+        await Promise.all(childs.map(c => this._elect({ ...data, nodeName: c })));
+    }
+
+    async _elect(options) {
+        const { jobId, nodeName } = options;
+        const key = `${jobId}/${nodeName}`;
+        const lock = await stateAdapter.acquireLock(key);
```
Comment (NassiHarel) [inline core/worker/lib/streaming/services/auto-scaler.js:50]: TODO:
release lock on finish

**Review by yehiyam — CHANGES_REQUESTED** [review]
Amazing work Nassi

> Reviewed 13 of 38 files at r1, 1 of 27 files at r3, 11 of 42 files at r5, 28 of 52 files at r6, 23 of 25 files at r7, 4 of 4 files at r8.
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/892#-:-MLJ8q7PDBxNPhtIDG0C:b-cyrkz5)* status: all files reviewed, 13 unresolved discussions (waiting on @golanha, @maty21, @NassiHarel, and @reggev)

___
*[core/api-server/package.json, line 9 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLIvMl69trmJvS-To37:-MLIvMl78bAH177sWY0P:bgejoex) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/api-server/package.json#L9)):*
> ```JSON
>         "@hkube/config": "^2.0.7",
>         "@hkube/consts": "^1.0.26",
>         "@hkube/dag": "^2.0.0",
> ```

don't forget to change to 2.1
___
*[core/pipeline-driver/package.json, line 13 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLIvTIM6ZnxIC9jZadl:-MLIvTIM6ZnxIC9jZadm:bgejoex) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/pipeline-driver/package.json#L13)):*
> ```JSON
>     "@hkube/config": "^2.0.7",
>     "@hkube/consts": "^1.0.29",
>     "@hkube/dag": "^2.0.13",
> ```

don't forget to change to 2.1
___
*[core/pipeline-driver/lib/tasks/task-runner.js, line 1 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLIun1pDd4r030eJbMM:-MLIun1q-EPJoy0t0oxL:bdq0vfj) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/pipeline-driver/lib/tasks/task-runner.js#L1)):*
> ```JavaScript
> const EventEmitter = require('events');
> ```

This file is more than 700 lines. I think it is good time to refactor it
___
*[core/worker/package.json, line 23 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLIvG06CMe25fnAq4k8:-MLIvG06CMe25fnAq4k9:bgejoex) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/package.json#L23)):*
> ```JSON
>         "@hkube/config": "^2.0.7",
>         "@hkube/consts": "^1.0.29",
>         "@hkube/dag": "^2.0.11",
> ```

don't forget to change to 2.1
___
*[core/worker/config/main/config.base.js, line 46 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLIwE8i5cPE1AO7SKhr:-MLIwE8i5cPE1AO7SKhs:bwjmbwp) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/config/main/config.base.js#L46)):*
> ```JavaScript
>         address: {
>             host: process.env.POD_IP || '127.0.0.1',
>             port: process.env.STREAMING_DISCOVERY_PORT || 9021
> ```

This is the same port as the ping port in the wrappers
___
*[core/worker/lib/states/stateAdapter.js, line 47 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLJ2fMTCeBqVoTrjQxm:-MLJ2fMTCeBqVoTrjQxn:b-lu0az4) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/lib/states/stateAdapter.js#L47)):*
> ```JavaScript
>             this.emit(res.status.command, res);
>         });
>         this.watch({ jobId: 'hookWatch' }); // should we really need it (we do watchWorkerStates)
> ```

probably not... Maybe even the mixer fixed this problem
___
*[core/worker/lib/streaming/adapters/adapter.js, line 6 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLJ3h5f1FQQR-YN70z0:-MLJ3h5f1FQQR-YN70z1:b-z8hjvy) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/lib/streaming/adapters/adapter.js#L6)):*
> ```
> isMaster
> ```
this is not politically correct :-)
___
*[core/worker/lib/streaming/adapters/adapters-proxy.js, line 10 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLJ3oMa1NNFdmU_X9mU:-MLJ3oMa1NNFdmU_X9mV:b-a9rppk) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/lib/streaming/adapters/adapters-proxy.js#L10)):*
> ```JavaScript
>  * with the adapters (master/slave)
>  */
> class AdaptersManager {
> ```

rename the file like the class
___
*[core/worker/lib/streaming/adapters/adapters-proxy.js, line 72 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLJ48BA96qdso1zbwwO:-MLJ48BA96qdso1zbwwP:b-7z18en) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/lib/streaming/adapters/adapters-proxy.js#L72)):*
> ```JavaScript
>     }
> 
>     getMasters() {
> ```

Is it possible to have more than one master?
___
*[core/worker/lib/streaming/core/interval.js, line 43 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLJ5C6j2lh7UaO8tqeJ:-MLJ5C6j2lh7UaO8tqeK:b1h90pc) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/lib/streaming/core/interval.js#L43)):*
> ```JavaScript
>     }
> 
>     onFunc(func) {
> ```

why `onFunc` and `onError` are not passed in the constructor?
___
*[core/worker/lib/streaming/core/throughput.js, line 10 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLJ5xUABuzNmKIlsTlI:-MLJ5xUABuzNmKIlsTlJ:bip4v59) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/lib/streaming/core/throughput.js#L10)):*
> ```
> this._data[nodeName] 
> ```
This is never cleared
___
*[core/worker/lib/streaming/services/auto-scaler.js, line 34 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLJ6HN76qR1z9OYPDUv:-MLJ6HN76qR1z9OYPDUw:b-5lpydz) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/lib/streaming/services/auto-scaler.js#L34)):*
> ```
> replicas * 1
> ```
this means no scale-up?

**Review by NassiHarel — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/892)* status: 56 of 71 files reviewed, 12 unresolved discussions (waiting on @golanha, @maty21, @reggev, and @yehiyam)

___
*[core/api-server/package.json, line 9 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLIvMl69trmJvS-To37:-MLMYiWe6d2CWXaFMjus:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/api-server/package.json#L9)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

don't forget to change to 2.1
</blockquote></details>

Done.
___
*[core/pipeline-driver/package.json, line 13 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLIvTIM6ZnxIC9jZadl:-MLMYj1LDvmgxOCxiBA0:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/pipeline-driver/package.json#L13)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

don't forget to change to 2.1
</blockquote></details>

Done.
___
*[core/pipeline-driver/lib/tasks/task-runner.js, line 1 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLIun1pDd4r030eJbMM:-MLMfGni8NXOYYBKYrtk:b-4oucbf) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/pipeline-driver/lib/tasks/task-runner.js#L1)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

This file is more than 700 lines. I think it is good time to refactor it
</blockquote></details>

I did a minor refactor.
___
*[core/worker/package.json, line 23 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLIvG06CMe25fnAq4k8:-MLMfPqQ4-EBj8jZY72l:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/package.json#L23)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

don't forget to change to 2.1
</blockquote></details>

Done.
___
*[core/worker/config/main/config.base.js, line 46 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLIwE8i5cPE1AO7SKhr:-MLMgv85BefbsYMT961k:b-m5xkgg) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/config/main/config.base.js#L46)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

This is the same port as the ping port in the wrappers
</blockquote></details>

This port was set before there was a ping :-) 
The ping port in the wrapper is (discovery port + 1) which is bad.
I will change it in worker and wrapper to 9022
___
*[core/worker/lib/states/stateAdapter.js, line 47 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLJ2fMTCeBqVoTrjQxm:-MLMhWLy8KlxSgzdfA5a:b-896fix) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/lib/states/stateAdapter.js#L47)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

probably not... Maybe even the mixer fixed this problem
</blockquote></details>

Done.
___
*[core/worker/lib/streaming/adapters/adapter.js, line 6 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLJ3h5f1FQQR-YN70z0:-MLMhqvE2l_h7TQH0e8s:b35fgau) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/lib/streaming/adapters/adapter.js#L6)):*
> ```
> this.watch({ jobId: 'hookWatch' })
> ```
Actually, in my case the Master does all the work, the slaves are doing nothing, so there is a small justice :-)
___
*[core/worker/lib/streaming/adapters/adapters-proxy.js, line 10 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLJ3oMa1NNFdmU_X9mU:-MLMi93U3kgHpeXj4AuR:b-vueg1i) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/lib/streaming/adapters/adapters-proxy.js#L10)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

rename the file like the class
</blockquote></details>

rename the class like the file
___
*[core/worker/lib/streaming/adapters/adapters-proxy.js, line 72 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLJ48BA96qdso1zbwwO:-MLMiIpl31VGaYJGtFcy:b2nhswy) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/lib/streaming/adapters/adapters-proxy.js#L72)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

Is it possible to have more than one master?
</blockquote></details>

Yes
___
*[core/worker/lib/streaming/core/interval.js, line 43 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLJ5C6j2lh7UaO8tqeJ:-MLMiKQn6JcxzDhhAkyT:b6016zq) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/lib/streaming/core/interval.js#L43)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

why `onFunc` and `onError` are not passed in the constructor?
</blockquote></details>

This is just a fluent pattern
___
*[core/worker/lib/streaming/core/throughput.js, line 10 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLJ5xUABuzNmKIlsTlI:-MLMisDn94H2SmvHwrze:b-5985cs) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/lib/streaming/core/throughput.js#L10)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

> ```
> this._data[nodeName] 
> ```
This is never cleared
</blockquote></details>


This is a new instance that created each time a master is created.
There are many more variables that never cleared like this, because almost every class in 
the streaming module is an instance (not singleton) that created and destroyed.
`stream-service.finish` should clean all these instances.
___
*[core/worker/lib/streaming/services/auto-scaler.js, line 34 at r8](https://reviewable.io/reviews/kube-hpc/hkube/892#-MLJ6HN76qR1z9OYPDUv:-MLMrHVW1_x40VlwtJmA:b2w0sn) ([raw file](https://github.com/kube-hpc/hkube/blob/76916b2d6f5ff728be5a0a176f3fa77e2427abf9/core/worker/lib/streaming/services/auto-scaler.js#L34)):*
<details><summary><i>Previously, yehiyam wrote…</i></summary><blockquote>

> ```
> replicas * 1
> ```
this means no scale-up?
</blockquote></details>

No, if there are only requests and no responses, the assumption is there are no receivers (first time) or there were receivers.
so if there is no resRate, I can't calc the reqResRatio to scale, the reqResRatio and currentSize will be 1 by default.  
so it will always calc (currentSize * ratio), just in this case it will be `currentSize(1) * 1`. it can be also `currentSize (10) * 1.`
I did a little change to not scale-up if there are only requests without responses and currentSize > 0.

**Review by yehiyam — APPROVED** [review]
:lgtm:

> Reviewed 15 of 15 files at r9.
> *[Reviewable](https://reviewable.io/reviews/kube-hpc/hkube/892#-:-MLN1A_C0ku4VW6bprd-:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @golanha, @maty21, and @reggev)

**Pull request merged**